rlay-backend-neo4j/src/lib.rs

```diff
@@ -220,76 +220,170 @@ impl Neo4jBackend {
+            // aka. those that are not self-referencing CIDs
+            for (key, value) in val.iter() {
                 if key == "cid" || key == "type" {
                     continue;
                 }
                 if (kind_name == "DataPropertyAssertion"
                     || kind_name == "NegativeDataPropertyAssertion")
                     && key == "target"
                 {
-                    values.push(format!("n.{0} = {1}", key, value));
                     continue;
                 }
                 if kind_name == "Annotation" && key == "value" {
-                    values.push(format!("n.{0} = {1}", key, value));
                     continue;
                 }
                 if let Value::Array(array_val) = value {
-                    for relationship_value in array_val {
-                        add_relationship_value(cid.clone(), key, relationship_value);
+                    for array_val_cid in array_val {
+                        relationships.push(RelationshipQueryPart{
+                            cid: array_val_cid.as_str().unwrap().to_string(),
+                            kind_name: key.clone()
+                        });
                     }
                     continue;
                 }
                 if let Value::String(_) = value {
-                    add_relationship_value(cid.clone(), key, value);
+                    relationships.push(RelationshipQueryPart{
+                        cid: value.as_str().unwrap().to_string(),
+                        kind_name: key.clone()
+                    });
                 }
             }
         }
 
-        let mut statement_query = format!(
-            "MERGE (n:RlayEntity {{cid: \"{1}\"}}) SET n:{0}",
-            kind_name, cid
+        let statement_query_main = format!("
+            MERGE (n:RlayEntity {{cid: $entity.cid }})
+            SET (CASE WHEN $entity.type = 'Annotation' THEN n END).value = $entity.value
+            SET (CASE WHEN $entity.type = 'DataPropertyAssertion' THEN n END).target = $entity.target
+            SET (CASE WHEN $entity.type = 'NegativeDataPropertyAssertion' THEN n END).target = $entity.target
+            WITH n
+            CALL apoc.create.addLabels(n, [$entity.type]) YIELD node
+            UNWIND $relationships as relationship
+                MERGE (m:RlayEntity {{ cid: relationship.cid }})
+                WITH n, m, relationship
+                CALL apoc.merge.relationship(n, relationship.kind_name, {{}}, {{}}, m) YIELD rel
+            RETURN n.cid",
         );
-        if !values.is_empty() {
-            statement_query.push_str(", ");
-            statement_query.push_str(&values.join(", "));
-        }
 
-        let (mut transaction, _) = client.transaction().begin().await?;
+        let statement_query = Statement::new(&statement_query_main)
+            .with_param("entity", &val)?
+            .with_param("relationships", &relationships)?;
```

Comment:
not sure if rusted cypher streams data to the server that's an option at least to write and read in a streaming manner